Commercial data API (#447)
Open-EO/openeo-api: extension/commercial_data/README.md

#### Example

```json
{
  "stac_version": "1.0.0",
  "stac_extensions": [
    "https://stac-extensions.github.io/datacube/v1.0.0/schema.json",
    "https://stac-extensions.github.io/order/v1.0.0/schema.json"
  ],
  "type": "Collection",
  "id": "PLEIADES",
  "title": "Airbus Pleiades",
  "description": "Pleiades is a satellite constellation providing very high-resolution optical imagery and is owned by Airbus. Pléiades is composed of two twin satellites orbiting the Earth 180° apart. The satellites deliver 0.5 m optical imagery and offer a daily revisit capability to any point on the globe.",
  "license": "proprietary",
  "providers": [...],
  "extent": {...},
  "cube:dimensions": {...},
  "summaries": {...},
  "assets": {...},
  "order:status": "orderable",
  "order_parameters": [
      {
        "name": "sample_type",
        "description": "sample type of the output raster.",
        "schema": {
          "type": "string"
        },

>         "enum": [

Reviewer: The enum should go into the schema, right?

Reviewer: Yeah, if `order_parameters` should basically be defined as `process_parameters` in `GET /service_types`, then you'd put the enum into the schema.

Author: Fixed.

          "UINT16",
          "UINT8"
        ]
      },
    ],
  "pricing": {
    "searching": null,
    "ordering": {
      "description": "Minimum area per order is 0.25 km2. The price is calculated based on a 6-months sliding window.",
    }
  },
  "links": [
    {
      "title": "Airbus Pleiades pricing",

>       "rel": "related",

Reviewer: See also https://github.com/Open-EO/openeo-api/pull/447#discussion_r924787502

>       "rel": "related",

Author: This was added.

```

#### `GET /orders/{order_id}`

Get full metadata of the order. The item should follow the [STAC Order Extension](https://github.com/stac-extensions/order), but extended with the spatial and temporal extent information and other metadata about the products.

```json
{
  "order:id": "40264b5-c3ae-46f4-a907-0f612d763d97",
  "order:status": "delivered",
  "order:date": "2017-01-01T09:32:12Z",
  "products": [
    {
      "type": "Feature",
      "stac_version": "1.0.0",
      "id": "c8a1f88d-89cf-4933-9118-45e9c1a5df20",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              12.36555287044679,
              41.94403289260048
            ]
          ],
          [
            12.36571746774068,
            41.86399361096971
          ],
          [
            12.60746759743069,
            41.86372776276345
          ],
          [
            12.60758647471871,
            41.94379931812686
          ],
          [
            12.36555287044679,
            41.94403289260048
          ]
        ]
      },
      "properties": {
        "constellation": "PHR",

>         "acquisitionDate": "2022-03-21T10:11:15.055Z",

Reviewer: ```suggestion
        "datetime": "2022-03-21T10:11:15.055Z",
```
That's how STAC defines and requires the acquisition date to be.

        "azimuthAngle": 179.9852862071639,

>         "cloudCover": 0,

Reviewer: Better use eo:cloud_cover

>         "geometryCentroid": {

Reviewer: Better use proj:centroid

Author: I changed all of this according to the suggestions, but I should clarify that this is an example of an Item as it is returned directly by Sentinel Hub, and that different products from different providers have various other fields, which are not present in this example.

Reviewer: Okay, then the spec should probably highlight that providers are recommended to align with STAC as much as possible and not only dump existing information "as-is" as then it's not very useful.

Author: Added a clarification.

          "lat": 41.903935647240964,
          "lon": 12.486569672582828
        },

>         "processingLevel": "SENSOR",

Reviewer: Better use processing:level

        "sensorType": "OPTICAL",
        "spectralRange": "VISIBLE"
      },
      "assets": {},
      "links": []
    }
  ],
  "parameters": {
    "bounds": [
      3,
      15,
      4,
      16
    ]
  },
  "collection_id": "PLEAIDES",

>   "costs": 42

Reviewer: This should go into properties, I think.

# Commercial Data Extension

The Commercial Data API extension provides an interface for discovering, ordering and using commercial data in the openEO API. 

- Version: **0.1.0**
- Stability: **experimental**

**Note:** This document only documents the additions to the specification.
Extensions can not change or break existing behavior of the openEO API.

## Overview of the workflow

> All the available datasets provided by a backend are listed on the `/collections` endpoint. The collections are normally freely accessible. This extensions adds capabilities for providing collections that are not free of charge and require purchasing data products that can thereupon be used in processing. Commercial data collections usually allow purchasing small subsets of the data (products), for example a single observation of an area.

Reviewer: ```suggestion
All the available datasets provided by a backend are listed on the `GET /collections` endpoint. The collections are normally freely accessible. This extensions adds capabilities for providing collections that are not free of charge and require purchasing data products that can thereupon be used in processing. Commercial data collections usually allow purchasing small subsets of the data (products), for example a single observation of an area.
```

Therefore, the client must have an ability to search the available products that match their desired criteria and inspect their metadata to decide whether the products should be purchased.

The client can then create an order for the desired products. Because of the financial cost of purchasing the data, a separate endpoint for confirming the execution of that specific order should be implemented.

> When the order is completed, the user should be able to access the data using the `collection-id` at the temporal and spatial location of the purchased products.

Reviewer: I think the parameter is generally called collection_id instead of collection-id in the API spec?!

```suggestion
When the order is completed, the user should be able to access the data using the `collection_id` at the temporal and spatial location of the purchased products.
```

### Collection discovery

> A backend should add general information about a commercial data collection to the `/collections` and `/collections/{collection-id}` endpoints, the same as with freely available collections. Only the metadata about the entire dataset needs to be provided, not about the specific data products that a user has already purchased. 

Author: I'm not sure this is needed, because the temporal and spatial availability of products is available in the information about orders.

However, this might be an argument for implementing `collections/{collection_id}/orders` as mentioned in https://github.com/Open-EO/openeo-api/pull/447#discussion_r924801576.

> Commercial data collections are distinguished from freely available collections by including `"order:status": "orderable"` as specified in [STAC Order specification](https://github.com/stac-extensions/order).

Reviewer: ```suggestion
Commercial data collections are distinguished from freely available collections by including `"order:status": "orderable"` as specified in the [STAC Order specification](https://github.com/stac-extensions/order).
```

Author: Thanks, fixed.

> Commercial data collections can include `order_parameters` field, if ordering supports additional parameters that specify how the products should be delivered.

Reviewer: ```suggestion
Commercial data collections can include an `order_parameters` field if ordering supports additional parameters that specify how the products should be delivered.
```

Author: Thanks, fixed.

  "stac_extensions": [
    "https://stac-extensions.github.io/datacube/v1.0.0/schema.json",
    "https://stac-extensions.github.io/order/v1.0.0/schema.json"
  ],
  "type": "Collection",
  "id": "PLEIADES",
  "title": "Airbus Pleiades",
  "description": "Pleiades is a satellite constellation providing very high-resolution optical imagery and is owned by Airbus. Pléiades is composed of two twin satellites orbiting the Earth 180° apart. The satellites deliver 0.5 m optical imagery and offer a daily revisit capability to any point on the globe.",
  "license": "proprietary",
  "providers": [...],
  "extent": {...},
  "links": [...],
  "cube:dimensions": {...},
  "summaries": {...},
  "assets": {...},
  "order:status": "orderable",
  "order_parameters": [
      {
        "name": "sample_type",
        "description": "sample type of the output raster.",
        "schema": {
          "type": "string"
        },
        "enum": [
          "UINT16",
          "UINT8"
        ]
      },
    ],
  "pricing": {
    "searching": null,
    "ordering": {
      "description": "Minimum area per order is 0.25 km2. The price is calculated based on a 6-months sliding window.",

>       "links": [

Reviewer: I think I'd propose to define relation types for this kind of links and instead add them to the links structure of the Collection.

Author: There is `payment`, but I that `Indicates a resource where payment is accepted.`.

It's strange that a relation type that just links to "additional information" does not exist. If we wanted to be more specific, I guess we could add `pricing-info | URL to human-readable pricing information.` or something like that?

        {
          "title": "Airbus Pleiades pricing",
          "rel": "related",
          "href": "https://www.sentinel-hub.com/pricing/#TPD_pricing"
        }
      ]
    }
  }
}
```

### Filtering parameters discovery

> Searching for products can support refining the search by filtering with general or collection-specific attributes. Backends should implement a top level `/queryables` endpoint for attributes available for all collections, and collection-specific attributes should be provided at `/collections/{collection-id}/queryables` according to [OGC Queryables specification](https://portal.ogc.org/files/96288#filter-queryables).

Reviewer: ```suggestion
Searching for products can support refining the search by filtering with general or collection-specific attributes. Backends should implement a top-level `/queryables` endpoint for attributes available for all collections, and collection-specific attributes should be provided at `/collections/{collection-id}/queryables` according to the [OGC Queryables specification](https://portal.ogc.org/files/96288#filter-queryables).
```

Author: Added.

#### Example

> Example response from `GET collections/PLEIADES/queryables`:

Reviewer: ```suggestion
Example response from `GET /collections/PLEIADES/queryables`:
```

Author: Thanks, fixed.

```json
{
   "$schema":"http://json-schema.org/draft-07/schema#",
   "$id":"http://example.com/commercial_data/PLEIADES/queryables",
   "type":"object",
   "title":"Queryables of Airbus Pleiades collection.",
   "description":"Available properties for CQL filtering of products.",
   "properties":{
      "processing_level":{
         "title":"Processing level",

>          "description":"When searching, you will receive results from the full catalog as well as the Living Library, which holds images that have cloud cover under 30% and Incidence angle under 40°. If you want to search only Living Library results, you will need to filter using processingLevel. This value could be equal to SENSOR (images which meet Living Library criteria) and ALBUM (images that do not meeting Living Library criteria in terms of incidence angle and cloud cover).'",

Reviewer: To get a more concise example, maybe shorten the text to a sentence?

Author: Done so.

         "type":"string",
         "enum":[
            "SENSOR",
            "ALBUM"
         ]
      },
      "max_snow_coverage":{
         "title":"Maximum snow coverage",
         "description":"The maximum allowable snow coverage in percent.",
         "type":"number",
         "minimum":0,
         "maximum":100,
         "default":100
      },
      "max_incidence_angle":{
         "title":"Maximum incidence angle",
         "description":"The maximum allowable incidence angle in degrees.",
         "minimum":0,
         "maximum":90,
         "default":90
      }
   }
}
```

### Searching available products

> Backends should implement top-level `GET /search` endpoint as specified in [STAC Item Search API specification](https://github.com/radiantearth/stac-api-spec/tree/v1.0.0-rc.1/item-search). This should include the [Filter Extension](https://github.com/radiantearth/stac-api-spec/tree/v1.0.0-rc.1/fragments/filter), which enables filtering the available products by attributes specified in `/queryables` and `collections/{collection-id}/queryables`. 

Reviewer: ```suggestion
Backends should implement the top-level `GET /search` endpoint as specified in the [STAC Item Search API specification](https://github.com/radiantearth/stac-api-spec/tree/v1.0.0-rc.1/item-search). This should include the [Filter Extension](https://github.com/radiantearth/stac-api-spec/tree/v1.0.0-rc.1/fragments/filter), which enables filtering the available products by attributes specified in `GET /queryables` and `GET /collections/{collection-id}/queryables`. 
```

Author: Thanks, fixed.

> The endpoint should return a list of  `STAC Item`s that match the criteria.

Reviewer: ```suggestion
The endpoint returns a list of STAC Items that match the criteria.
```

Author: Thanks, fixed.

#### Example

Example request payload to `GET /search` for `PLEIADES` products from "Living Library" with no snow coverage.

> ```json

Author: Did the same for some others as well.

{
  "bbox": [
    3,
    15,
    4,
    16
  ],
  "datetime": [
    "01-01-2022",
    "01-02-2022"
  ],

>   "collectionsArray": [

Reviewer: This parameter is called collections in STAC.

```suggestion
  "collections": [
```

    "PLEIADES"
  ],
  "filter": {
    "op": "and",
    "args": [
      {
        "op": "=",
        "args": [
          {
            "property": "processing_level"
          },
          "SENSOR"
        ]
      },
      {
        "op": "=",
        "args": [
          {
            "property": "max_snow_coverage"
          },
          0
        ]
      }
    ]
  }
}
```

### Ordering products

Backends should implement the following endpoints:

> - `GET /orders`: Get a list of all created orders

Reviewer: I'm still wondering whether all endpoints should generally live under `/collections/{collection_id}/orders` with an additional `GET /orders` that lists all orders with limited metadata. Not sure yet, but posting it here for discussion.

Reviewer: Resolving this in favor of https://github.com/Open-EO/openeo-api/pull/447#discussion_r924812905

- `POST /orders`: Create an order

> - `GET /orders/{order-id}`: Get full metadata of a specific order

Reviewer: The parameter should be order_id instead of order-id, I think.

    links:
      $ref: 'https://raw.githubusercontent.com/Open-EO/openeo-api/1.0.0/openapi.yaml#/components/schemas/links_pagination'
```

#### `POST /orders`

Create an order for selected products. Order can contain some additional parameters that specify how the products should be delivered. For example, depending on the collection it might be possible to set the projection, resampling method, bit depth etc of the delivered data.

Backends should expose the available ordering parameters in `/collections/{collection-id}` in the `order_parameters` field, following the `process_parameters` schema of [`GET /service_types`](https://docs.openeo.cloud/api/#tag/Secondary-Services/operation/list-service-types). 

The request should return the `order-id`, current status and the costs of the order.

```yaml
schema:
  title: Create Order Request
  type: object
  required:
    - collection_id
    - products
    - parameters
  properties:
    collection_id:
      type: string
      description: Unique identifier of the collection.
    products:
      type: array
      description: Array of IDs of products to order.
      items:

>         $ref: 'https://raw.githubusercontent.com/Open-EO/openeo-api/1.0.0/openapi.yaml#/components/schemas/order_id'

Reviewer: In the OpenAPI document, you should better use relative links.

    parameters:
      type: object
      description: Key-value pairs of available `order_parameters` as listed at `GET /commercial_data/collections/{collection_id}` for filtering available products.
      additionalProperties:
        x-additionalPropertiesName: Order Parameter Name
        description: Value of the order parameter to be used in the request.
```

#### `GET /orders/{order-id}`

> Get full metadata of the order. The item should follow the [STAC Order Extension](https://github.com/stac-extensions/order), but extended with the spatial and temporal extent information and other metadata about the products.

Reviewer: I'm wondering whether this should actually return a STAC Collection?

```json
{
  "order:id": "40264b5-c3ae-46f4-a907-0f612d763d97",
  "order:status": "delivered",
  "order:date": "2017-01-01T09:32:12Z",

>   "products": [

Reviewer: Instead of inlining them, we may want to consider linking to them. Also, I'm wondering whether we should generally also think about which of our additions we can propose to be added to the STAC extension?

Reviewer: Thinking more about it, we really should list the products at /collections/{collection_id}/items and just link to them from here so that we don't need to define the whole STAC structure again for this endpoint.

Reviewer: See also https://github.com/Open-EO/openeo-api/pull/447#discussion_r924801576

Author: This would be quite a problem for our implementation, as Sentinel Hub doesn't support accessing information about a specific product directly, or listing all of them for that matter. It only supports searching them (which behind the scenes uses the searching functionality of the third parties who provide the data). 

In the "previous" proposal, `GET /orders` did not include any "heavy" information  about products in order to make it lightweight: https://github.com/Open-EO/openeo-api/pull/447/commits/27a806cc44ddb4e8bc1551c37b995464f6569046. This was intended in this proposal as well, perhaps I should clarify that. The products are listed inline only in `GET /orders/{order_id}`.

I would suggest we add `/items` and linking them as optional, and allow listing them inline by default.

Reviewer: But isn't listing a list of products for a specific collection basically the same as searching all products for a given collection ID? The same applies for the Item ID, I'd think.

Reviewer 2: While I agree with @m-mohr that we should aim to reuse the existing STAC interfaces as much as possible, our implementation will inevitably have to take into account some limitations, specifically:

When listing `orderable` items:
- listing collection items for non-ordered products is only possible using Sentinel Hub's [3rd party data searching request](https://docs.sentinel-hub.com/api/latest/reference/#operation/dataImport_searchData);
- the generic part of the simple search API is _very_ limited (the only query parameters one can use in a generic way are `bounds`, `timeRange` and `maxCloudCoverage`);
- there are more options in the provider-specific "native search", but even these **don't currently allow fetching products by their ID** (except maybe with Planet data, where the native search is the most sophisticated);
- you always get all item data inline; there's paging to make 100+ products use-case manageable.

When listing `ordered` items:
- we can list all `product IDs` that were used as input when creating [a specific order](https://docs.sentinel-hub.com/api/latest/reference/#operation/dataImport_getOrder)
- after delivery, the delivery-related [endpoints](https://docs.sentinel-hub.com/api/latest/reference/#operation/dataImport_getOrderDelivery) again list only files and product IDs, but they don't contain STAC item metadata for the delivered products
- STAC item metadata is only available when the images are ingested into user collections (BYOC) within Sentinel Hub, and accessible via the specific collection's [catalog API](https://docs.sentinel-hub.com/api/latest/data/byoc/#catalog-api-capabilities)

My hope is that we can come up with a commercial data extension that is both reasonable as a standard interface as well as not-too-difficult to implement with the above constraints.

Author: > the generic part of the simple search API is very limited (the only query parameters one can use in a generic way are bounds, timeRange and maxCloudCoverage);

We already have `/search` in this spec, so I think the searching functionality of `/items` isn't really needed, and I also don't see a big utility in listing all available products.

IMO the only really useful thing is to have information about a single item on a separate endpoint, but as you mentioned this is exactly what we don't support atm. That's why I proposed to have it optional.

What we could also to is to have e.g. `GET /orders/{order_id}/products` or `GET /orders/{order_id}/items` (with pagination) and list all the products  metadata there.

Reviewer: Given the limitations, I have two ideas:
1. Treat orders as new collections, i.e. each order creates a new collection at /collections and has then additional metadata for this specific order. You can then load specific orders via load_collection etc, but may end up with a lot of collections. You could then access the new data in two ways:
    1. via the original collection where the data for all orders can be accessed via load_collection.
    2. via the new collection where only data for the specific order can be accessed via load_collection
2. Do the same as above, but keep the new collections separate at /orders. Still, they should be full collections so that we can re-use as much code as possible from the collection implementations in the clients and back-ends, they just have some additional order-related metadata included.

Whether to choose 1 or 2 is more a "strategic" question on whether we want new orders to pop up in the collections endpoint or not.

Generally (i.e. for 1 and 2), no product metadata is inlined and you access them via .../items. Back-ends are free to choose whether they can make that only available for the new collections or also the old ones. You can give some information about products in that order, e.g. by listing product ids in collection summaries or so.

I think that should work having the limitations in mind, but maybe I missed something.

Author: > You could then access the new data in two ways:
> i. via the original collection where the data for all orders can be accessed via load_collection.
> ii. via the new collection where only data for the specific order can be accessed via load_collection

Doing something like this, i.e. saving purchased data to separate collections but allow loading them as one, is possible for us, as we have a "data fusion" feature, but if I'm not mistaken, requests to i.) would become linearly more expensive as new orders are added, so I don't think it's preferable (and it's definitely more complex). 

IMO it would be preferably to:

1. Have a single orderable "general" collection that contains data for all the orders and is used in `load_collection`, e.g. `collections/PLEIADES`
2. (Orderable) collections have `/items` where all ordered items are listed, e.g. `collections/PLEIADES/items`
3. I guess we will have to save the product Items on our backend and implement `collections/PLEIADES/items/{item-id}`, but only for the products used in orders, and not generally.
4. `orders/{order-id}` then use links to items

> I think that should work having the limitations in mind, but maybe I missed something. Happy to jump on another quick call to discuss details.

I would be happy to, when do you have time?

Reviewer: Today 13:00, 16:00, tomorrow 11:00, 13:00, > 15:00, Fri 13:00

Author: Today at 16:00 is fine for me, I think you should have received the invite.

Author: Sorry about the meeting time limit, it's maximum by Webex on my account.

> Retrieve order information: GET /orders/:id -> contains metadata about the order (source and target collection id [might be the same or different], order extension fields, optionally links to product metadata), no inline product metadata

+ array of ordered `product-id`s

> Optional: Retrieve product metadata: GET /collections/:id/items and GET /collections/:id/items/:item_id -> may return error if not ingested yet, otherwise list of items. The items should contain links to the respective orders that made them available.

Ok let's go with this, as it makes the most sense.
My problem is just the Sentinel Hub implementation, as I don't think it is possible for items in `GET /collections/:id/items` to equal the ones the user queried with `/search` and ordered, with existing Sentinel Hub services (`GET /collections/:id/items` would contain only basic information about spatial and temporal extent).

However, this solves the issue of linking orders to items very nicely.

Reviewer 2: > Can orders (in the future?) be created for multiple collections? e.g. ordering PLANET and MAXAR data at once?

I don't see this as a high priority requirement - cross-provider search might be useful sometimes, but cross-provider ordering seems too complicated to implement for not-so-much gain.

The interface would not need to change much, though - the `/order` endpoint is top-level, so we'd just need to include `collection_id` per each item and not have just one separate collection_id parameter. And we'd have to support multiple-output-collections-per-order. But I still suggest to leave this for some future upgrade.

Reviewer: The question about cross-provider orders was actually related to the question whether /orders or /collections/:id/orders would be preferrable. And I now used /orders above to have exactly this flexibility, otherwise we could also go with /collections/:id/orders, I think

Reviewer 2: > ... we could also go with /collections/:id/orders, I think

Even then, I think it's better to have `/orders`, so that all the user's orders are available in one place via  `GET /orders`. If one would want just a subset, we could at some point add filtering to this endpoint.

Author: I updated the README.md and I added the openAPI file for `/orders` and `/orders/{order_id}` endpoints. Should I add `/search`, `/items`, `queryables` etc. as well? Sadly it's not possible to reference them directly from Github because they contain relative references, so I'd have to copy most of it over.

Reviewer: I'll have a look next week, thanks. I don't think we necessarily $ref the other endpoints as long as they are mentioned (and linked to) in the README.

      "id": "c8a1f88d-89cf-4933-9118-45e9c1a5df20",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              12.36555287044679,
              41.94403289260048
            ]
          ],
          [
            12.36571746774068,
            41.86399361096971
          ],
          [
            12.60746759743069,
            41.86372776276345
          ],
          [
            12.60758647471871,
            41.94379931812686
          ],
          [
            12.36555287044679,
            41.94403289260048
          ]
        ]
      },
      "properties": {
        "constellation": "PHR",
        "acquisitionDate": "2022-03-21T10:11:15.055Z",
        "azimuthAngle": 179.9852862071639,
        "cloudCover": 0,
        "geometryCentroid": {
          "lat": 41.903935647240964,
          "lon": 12.486569672582828
        },
        "processingLevel": "SENSOR",
        "sensorType": "OPTICAL",
        "spectralRange": "VISIBLE"
      },
      "assets": {},
      "links": []
    }
  ],
  "parameters": {
    "bounds": [
      3,
      15,
      4,
      16
    ]
  },
  "collection_id": "PLEAIDES",
  "costs": 42
}
```

> #### `POST /orders/{order-id}`

Reviewer: What happens if I call this twice or so?

When an order is created, the data isn't yet ordered from the commercial data provider. The client should explicitly confirm the order whereupon the order is executed, the costs are deducted from the client's account and the data is ingested in the associated collection. The `order:status` changes to `ordered`.

If the user doesn't have sufficient funds, the endpoint should return an error and `order:status` should not change.

### Payment

Payment should be done in the currency used by the backend, listed at `GET /` under `billing`. When an order is created, the backend should return the full final cost of the order. 

> ### Example usage with Python client

Reviewer: I haven't closely checked the Python example yet, I think @soxofaan or another Python client dev should also have a look.

    {
      "type": "Feature",
      "stac_version": "1.0.0",

>       "stac_extensions": ["proj", "eo", "view", "processing"],

Reviewer: These are the v0.9 extension names, in v1.0.0 you need the schema URIs. For example `https://stac-extensions.github.io/eo/v1.0.0/schema.json` instead of `eo`. Please check the extension READMEs for the corresponding URIs.

Author: Thanks, fixed.
